src/cargo/core/compiler/custom_build.rs

```diff
@@ -249,6 +249,28 @@ fn build_work(cx: &mut Context<'_, '_>, unit: &Unit) -> CargoResult<Job> {
         }
     }
 
+    // Also inform the build script of the rustc compiler context.
+    if let Some(wrapper) = bcx.rustc().wrapper.as_ref() {
+        cmd.env("RUSTC_WRAPPER", wrapper);
+    }
+    cmd.env("RUSTFLAGS", bcx.rustflags_args(unit).join(" "));
```

Comment:
A downside of this that I'm now realizing is that this is a lossy representation of `RUSTFLAGS` due to the `.join(" ")` for when arguments have spaces in them.

That being said I don't know of a great way to fix this. A fix could be to export `RUSTFLAGS_JSON` as well with a JSON-encoding, but it's not like we want all build scripts to depend on a json decoder as well.

The other solution is we just say "whelp sorry" and don't support that configuration via `RUSTFLAGS` communicated to build scripts, and while Cargo will work with spaces-in-args (if you specify via TOML) build scripts that require reading RUSTFLAGS then won't work.

I don't really know the right tradeoff here.

Reply:
I definitely don't want to add dependencies to `autocfg`. I could probably hand-code something relatively simple like shell-quoting though.

Reply:
(I do think it needs a distinct env name though, if there's any kind of escaping introduced.)

Reply:
Well this is my point about I don't know a great way to fix this. I don't expect anyone to take on a json dependency to fix this, and something like shell escaping brings up weird questions like "what syntax do we use?" Like should we shell-escape based on the system? There's no reason for shell escaping anyway since this is rarely going to make its way into a shell anyway.

I really don't know of a great way to fix this. As-is it's basically saying "sorry, if you have a space in your path I hope you don't have a build script relying on that", which doesn't seem great. It seems slightly better than the state today, though, where none of this information is present at all.

Reply:
Would it make sense to use one of the ASCII separators?
https://en.wikipedia.org/wiki/C0_and_C1_control_codes#Field_separators

I don't know anyone that does this for real, but I wonder if that's just because they're not visible characters. At least they're unlikely to collide with normal use cases.

Reply:
This page does mention using US like that: https://en.wikipedia.org/wiki/Delimiter#ASCII_delimited_text

Reply:
Actually, I worry that escaping won't actually work, because when we then make the next trip into cargo, it'll parse `RUSTFLAGS` by simple whitespace splitting anyway:
https://github.com/rust-lang/cargo/blob/0cecbd67323ca14a7eb6505900d0d7307b00355b/src/cargo/core/compiler/build_context/target_info.rs#L598-L604

I realized this while trying to write a test-case for escaping. Consider a `.cargo/config.toml` with:
```toml
[build]
rustflags = ["-Clink-arg=-add_empty_section text foobar"]
```

With escaping, `RUSTFLAGS` would be set to `'-Clink-arg=-add_empty_section text foobar'` (note the quotes), which is all good and well. But if the build script that's passed that then invokes Cargo, Cargo will use its naive whitespace splitting linked above, and invoke `rustc` with

 - `'-Clink-arg=-add_empty_section`
 - `text`
 - `foobar'`

Is that a bug in Cargo? Maybe. Can we fix that? Probably not without breaking existing callers.

Not sure what we do about this one. Given the incompatibility, I think there's a decent argument for exposing the escaped form as `CARGO_RUSTFLAGS` to explicitly denote that its not compatible with `RUSTFLAGS`.. Or maybe `CARGO_ESCAPED_RUSTFLAGS`?

Reply:
Sorry again for not realizing this sooner, it really does complicate things :(

I agree, though, that if setting `RUSTFLAGS` *would* be lossy then we should not set it, rather it should be forcibly unset. If we do some encoding scheme, whatever it is, then it should go into a separate environment variable (probably something like `CARGO_ESCAPED_RUSTFLAGS` or `ENCODED` depending on what's implemented). W/e we set for build scripts Cargo would also want to be updated to recognize and read it as well (and we'd want to document it).

AFAIK ever other build system's approach to this is either:

* Don't uses spaces
* We're using a shell internally so use shell escaping

Basically I do think that we have some leniency of what to do here. Given the relative rarity of spaces-in-paths I would still lean towards setting `RUSTFLAGS` if it's lossless, although this could create surprising behavior if `RUSTFLAGS` isn't set just because you have a space somewhere and some build script suddenly breaks because it didn't account for that. I don't think I lean too much in this direction...

In terms of encoding scheme though if we can do some ascii trick that basically no one else uses that seems fine by me. I think it may be the first-of-its-kind build-system-wise but this is all purely-machine-consumed stuff that we can clearly document as well. I've never seen anything else use 0x1f in my life and presumably it's a whole lot rarer than spaces-in-paths so it seems like a reasonable thing to use.

Reply:
I like using 0x1F for this actually — it's not as easy to construct by hand, but it's also less error-prone than trying to do shell escaping from memory. I'm a little hesitant to have Cargo pick up this value automatically because it means it'll have to pick a precedence between the two, which may be confusing to users, but maybe we can just issue a warning if both are set?

So, how about this:

 - Cargo sets `ENCODED_RUSTFLAGS` to be `bcx.rustflags().join(0x1f)` for build scripts. It also unsets `RUSTFLAGS`.
 - If `ENCODED_RUSTFLAGS` is set, Cargo uses it to compute the rust flags to pass by splitting on `0x1f`.
 - If `ENCODED_RUSTLFAGS` and `RUSTFLAGS` are both set, Cargo issues a warning and uses `ENCODED_RUSTFLAGS`

Reply:
Bikeshed-wise I'd like to prefix the env var name with `CARGO_` since it's Cargo-specific. Otherwise though I don't think that a warning needs to be issued and we can just defined that the encoded version takes precedent. The encoded version is never expected to be set by humans and so I don't think it'll really come up often that both are set. This I think otherwise runs the risk of generating annoying warnings that are hard to squelch.

Reply:
Bikeshed: isn't `RUSTFLAGS` just as Cargo-specific?

Reply:
Pushed a commit with `CARGO_ENCODED_RUSTFLAGS`.